# Algoritmo Genetico

In [1]:
import numpy as np

Construimos la funcion objetivo, usa la funcion Rosenbrock.

In [2]:
#Rosenbrock
fob = lambda x: -1*(x[0]**4+3*x[0]**3*x[1]+\
                6*x[0]**2*x[1]**2+3*x[0]*x[1]**3+\
                x[1]**4-4*(x[0]-x[1])*(x[0]+3*x[1]))

Codigo del algoritmo genetivo visto en la diapositivas.

In [3]:
#Generar poblacion
N = 100
P = 2
X0 = 4*np.random.rand(N,P)-2 #Poblacion inicial
X = X0.copy()
#Evaluar poblacion
aptitud = []
for k in range(N):
    aptitud.append(fob(X[k,:]))

aptitud = np.array(aptitud)
indSort = np.argsort(aptitud)

print(max(aptitud), min(aptitud))

for k in range(100):
  #Ruleta
  ValoresRul = aptitud - aptitud.min() + 0.1
  S = sum(ValoresRul)
  r = np.random.rand()*S
  sumasPar = np.cumsum(ValoresRul)

  ind = np.argwhere(sumasPar>=r)[0][0]


  sinInd = list(range(N))
  sinInd.remove(ind)
  S2 = sum(ValoresRul[sinInd])
  r = np.random.rand()*S2
  sumasPar = np.cumsum(ValoresRul[sinInd])

  ind2=np.argwhere(sumasPar>=r)[0][0]


  #cruce
  alpha = np.random.rand()
  hijo = alpha*X[ind,:]+(1-alpha)*X[ind2,:]

  #Mutacion
  if np.random.rand()>0.5:
    cuantos_par = np.random.randint(X.shape[1]+1)
    indHijoMutar = np.random.permutation(X.shape[1])[:cuantos_par]
    hijo[indHijoMutar] += np.random.randn(cuantos_par)

  #Evolucion
  X[indSort[0],:] = hijo
  aptitud[indSort[0]] = fob(hijo)
  indSort = np.argsort(aptitud)

bestInd = np.argmax(aptitud)
mejorX = X[bestInd,:]
mejorAptitud = aptitud[bestInd]
print("La solucion es ",mejorX," con aptitud: ",mejorAptitud)

3.201375272005928 -188.17969086297322
La solucion es  [-1.15207517  0.05959591]  con aptitud:  3.201375272005928


Buscamos en una grilla (grid) el optimo.

In [4]:
x1 = np.linspace(-2,2,100)
x2 = np.linspace(-2,2,100)
x1v, x2v = np.meshgrid(x1,x2)

In [5]:
A = np.hstack((x1v.reshape(100*100,1),x2v.reshape(100*100,1)))

In [6]:
aptitudFB = []
for i in range(A.shape[0]):
  aptitudFB.append(fob(A[i,:]))
aptitudFB = np.asarray(aptitudFB)

In [7]:
np.max(aptitudFB),A[np.argmax(aptitudFB),:]

(np.float64(4.092306317199881), array([-1.39393939, -0.06060606]))

Segun la grilla, el maximo es 4.09 ubicado en $x_1=-1.39$, $x_2=-0.061$.

# PSO - Enjambre de Particulas

In [8]:
#Este codigo del PSO esta hecho para maximizar
Particulas = X0.copy() #misma poblacion inicial del genetico

l_inf = np.array([-2.,-2.])
l_sup = np.array([2.,2.])

# Ejemplo para minimizacion
#funobj = lambda x: np.sum(x**2) #Funcion a minimizar
#fob = lambda x: -funobj(x) #Se invierte para maximizar

F = np.apply_along_axis(fob,1,Particulas)

pbest = Particulas.copy()
Fpbest = F.copy()
indBest = np.argmax(F)
gbest = Particulas[indBest,:].copy()
Fgbest = F[indBest]
Iteraciones = 100
c1, c2 = 0.5, 0.5
Velocidad = np.zeros_like(Particulas)
for i in range(Iteraciones):
  r1 = np.random.rand(1,P)
  r2 = np.random.rand(1,P)
  #Calculo de la velocidad para todas las particulas
  Velocidad += c1*r1*(pbest-Particulas)+c2*r2*(gbest-Particulas)
  #Actualizacion posicion de todas las particulas
  Particulas += Velocidad
  Particulas = np.clip(Particulas,l_inf,l_sup)
  F = []
  for k in range(N): #Evaluamos fob para cada particula
    F.append(fob(Particulas[k,:]))
    if (F[k]>Fpbest[k]): #Actualizamos el mejor local
      Fpbest[k] = F[k]
      pbest[k,:] = Particulas[k,:].copy()

  indBest = np.argmax(F)
  if (F[indBest]>Fgbest): #Actualizamos el mejor global
    Fgbest = F[indBest]
    gbest = Particulas[indBest,:]

print('x: ',gbest,', f(x): ',fob(gbest))

x:  [-1.0955686   0.39246423] , f(x):  -1.3137252127960082


In [9]:
A = 5*np.random.rand(10,2)-2

In [10]:
Ap = np.clip(A,[-2,-1],[1,2])

In [11]:
A,Ap

(array([[ 2.31545311, -0.84544341],
        [-0.70710658, -0.10782869],
        [ 2.17822859, -1.82618035],
        [ 0.65903991,  0.44302678],
        [-1.17831809, -0.73934708],
        [ 1.35875493,  0.8375633 ],
        [ 2.43352427,  2.62250537],
        [-0.08838097, -0.08596948],
        [ 1.131886  , -0.34922415],
        [-1.21023195, -0.7764524 ]]),
 array([[ 1.        , -0.84544341],
        [-0.70710658, -0.10782869],
        [ 1.        , -1.        ],
        [ 0.65903991,  0.44302678],
        [-1.17831809, -0.73934708],
        [ 1.        ,  0.8375633 ],
        [ 1.        ,  2.        ],
        [-0.08838097, -0.08596948],
        [ 1.        , -0.34922415],
        [-1.21023195, -0.7764524 ]]))